identity-aware-proxy/IAPClient/IAPClient.cs

```diff
@@ -0,0 +1,129 @@
+            using (var fs = new FileStream(credentialsPath,
+                FileMode.Open, FileAccess.Read))
+            {
+                credentials = NewtonsoftJsonSerializer.Instance
+                    .Deserialize<Credentials>(fs);
+            }
+
+            // Generate a JWT signed with the service account's private key 
+            // containing a special "target_audience" claim.
+            var jwtBasedAccessToken =
+                CreateAccessToken(credentials.PrivateKey, iapClientId,
+                    credentials.ClientEmail);
+
+            var body = new Dictionary<string, string>
+            {
+                { "assertion", jwtBasedAccessToken },
+                { "grant_type", "urn:ietf:params:oauth:grant-type:jwt-bearer"}
+            };
+
+            // Request an OIDC token for the Cloud IAP-secured client ID.
+            var httpClient = new HttpClient();
+            var httpContent = new FormUrlEncodedContent(body);
+            var result = httpClient.PostAsync(GoogleAuthConsts.OidcTokenUrl,
+                httpContent).Result;
+            var responseContent = result.Content.ReadAsStringAsync().Result;
+            int statusCode = (int)result.StatusCode;
+            if (statusCode < 200 || statusCode >= 300)
+            {
+                throw new HttpRequestException(string.Format("{0} {1}\n{2}",
+                    statusCode, result.ReasonPhrase, responseContent));
+            }
+            string token = JsonConvert.DeserializeObject<IapResponse>(
+                responseContent).IdToken;
+
+            // Include the OIDC token in an Authorization: Bearer header to 
+            // IAP-secured resource
+            httpClient.DefaultRequestHeaders.Authorization =
+                new AuthenticationHeaderValue("Bearer", token);
+            string response = httpClient.GetStringAsync(uri).Result;
+            return response;
+        }
+
+        static long ToUnixEpochDate(DateTime date)
+              => (long)Math.Round((date.ToUniversalTime() -
+                                   new DateTimeOffset(1970, 1, 1, 0, 0, 0,
+                                        TimeSpan.Zero)).TotalSeconds);
+
+        static string CreateAccessToken(string privateKey,
+            string iapClientId, string email)
+        {
+            var now = DateTime.UtcNow;
+            var currentTime = ToUnixEpochDate(now);
+            var expTime = ToUnixEpochDate(now.AddHours(1));
+
+            var claims = new[]
+            {
+                new Claim(JwtRegisteredClaimNames.Aud,
+                    GoogleAuthConsts.OidcTokenUrl),
+                new Claim(JwtRegisteredClaimNames.Sub, email),
+                new Claim(JwtRegisteredClaimNames.Iat, currentTime.ToString()),
+                new Claim(JwtRegisteredClaimNames.Exp, expTime.ToString()),
+                new Claim(JwtRegisteredClaimNames.Iss, email),
+
+                new Claim("target_audience", iapClientId)
```

Comment:
Please, document this line (target_audience). This is the main reason you need to create a signature. People should understand what are you doing here.

Reply:
What should the comment say?  The only docs I find that describe target_audience are here:
https://cloud.google.com/iap/docs/authentication-howto
And they don't say much about it.

Reply:
We need to generate a JWT signed with the service account's private key containing a special "target_audience" claim. That claim should contain the clientId of IAP we eventually want to access.

You can take a look at the same page you mentioned but open python sample. It's better documented than Java.

Reply:
Added.